packages/lib-sourcify/src/Verification/Transformations.ts

```diff
@@ -0,0 +1,348 @@
+import { AuxdataStyle } from '@ethereum-sourcify/bytecode-utils';
+import { ImmutableReferences } from '../Compilation/SolidityTypes';
+import {
+  CompiledContractCborAuxdata,
+  LinkReferences,
+  Metadata,
+  StringMap,
+} from '../Compilation/CompilationTypes';
+import { AbiConstructor } from 'abitype';
+import { defaultAbiCoder as abiCoder, ParamType } from '@ethersproject/abi';
+import { id as keccak256Str } from 'ethers';
+import { logError } from '../logger';
+
+export type Transformation = {
+  type: 'insert' | 'replace';
+  reason:
+    | 'constructorArguments'
+    | 'library'
+    | 'immutable'
+    | 'cborAuxdata'
+    | 'callProtection';
+  offset: number;
+  id?: string;
+};
+
+// Call protection is always at the start of the runtime bytecode
+export const CallProtectionTransformation = (): Transformation => ({
+  type: 'replace',
+  reason: 'callProtection',
+  offset: 1, // 1 byte is always the PUSH20 opcode 0x73
+});
+
+// TransformationValues only has one ConstructorTransformatino so no id field is needed
+export const ConstructorTransformation = (offset: number): Transformation => ({
+  type: 'insert',
+  reason: 'constructorArguments',
+  offset,
+});
+
+export const AuxdataTransformation = (
+  offset: number,
+  id: string,
+): Transformation => ({
+  type: 'replace',
+  reason: 'cborAuxdata',
+  offset,
+  id,
+});
+
+export const LibraryTransformation = (
+  offset: number,
+  id: string,
+): Transformation => ({
+  type: 'replace',
+  reason: 'library',
+  offset,
+  id,
+});
+
+export const ImmutablesTransformation = (
+  offset: number,
+  id: string,
+  type: 'replace' | 'insert',
+): Transformation => ({
+  type,
+  reason: 'immutable',
+  offset,
+  id,
+});
+
+export interface TransformationValues {
+  constructorArguments?: string;
+  callProtection?: string;
+  libraries?: {
+    [id: string]: string;
+  };
+  immutables?: {
+    [id: string]: string;
+  };
+  cborAuxdata?: {
+    [id: string]: string;
+  };
+}
+
+// returns the full bytecode with the call protection replaced with the real address
+export function checkAndCreateCallProtectionTransformation(
```

Comment:
Maybe rename these all functions `transformCallProtection` etc. ?

Reply:
Ok I see and agree. Shall we say something like `extractConstructorArgsTransformation` instead of `checkAndCreate...` or do you want to keep that part too?